In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
#Establish Path
path = r'C:\Users\alber\Instacart Basket Analysis'

In [57]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [8]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [12]:
dfords.describe()
# order_days_of_week shouldn't have a maximum larger than '7' & the 50th percentile should be '3'
# days_since_prior_order should not be a negative minimum value
# order_hour_of_day should not have a maximum larger than '24', since there are only 24 hours in a day

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_days_of_week  order_hour_of_day  days_from_previous_order  \
count        3.421083e+06       3.421083e+06              3.214874e+06   
mean         2.776219e+00       1.345202e+01              1.111484e+01   
std          2.046829e+00       4.226088e+00              9.206737e+00   
min          0.000000e+00       0.000000e+00              0.000000e+00   
25%          1.000000e+00       1.000000e+01              4.000000e+00   
50%          3.000000e+00       1.300000e+01              7.000000e+00   
75%          5.000000e+00       1.600000e+01              1.500000e+01   
max          6.000000e+00       2.300000e+01              3.000000e+01   

       orer_hour_of_day  
count      3.421083e+06  
mean       1.345202e+01  
std        4.226088e+00  
min        0.000000e+00  
25%        1.000000e+01  
50%        1.300000e+01  
75%        1.600000e+01  
max        2.300000e+01

In [14]:
# Create a dataframe

df_test = pd.DataFrame()

In [15]:
# Create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [16]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [23]:
# Check for mixed types

for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print (col)
        
# Tab spacing is important
# weird (new variable) tests whether data types within column are consistent
# 'if' statement is a condition
# if werid is True, print(col) command is executed, which prints the problematic column for you to see.

mix


In [24]:
# Deciding what single data type the column in question should be
# If names, should be string
# If numbers, should be numeric
# Once decided, run following code.

df_test['mix'] = df_test['mix'].astype('str')

# Converts data type from numeric to string, but sometimes you will need to go from string to numeric.

In [27]:
## Missing Values
# Can occur from 1) data corruption, 2) unreccorded values

df_prods.isnull().sum()

# Assigns isnull() to df_prods, then sums the result with sum() function.
# isnull() used to find missing observations with 'observations,' referring to entries in dataframe

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [29]:
# We can see that product_name is missing 16 values. 
# To see these 16 values, create subset of dataframe only containing these values.

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [30]:
df_nan

# NaN = Not A Number (Missing Value)

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [31]:
# Few Ways to Deal with Missing Data:
# 1. Create new variable that acts like a flag based on missing value
    # Ex. Creating Column containing string values 'Delinquent' and 'Not Delinquent'
    # Important to understand the data you are working with
    
# 2. Impute the value with the mean or median of column (if variable is numeric)
    # Usually the median/mean.
    # Mean could be greatly influenced by extreme values (outliers)
    # If using mean, can use df.describe() to find mean of column
        # df['column with missings'].fillna(mean value, inplace = True)
    # IF using median, can use df_prods.median() to find median
        # df['column with missings'].fillna(median value, inplace = True)
        
# 3. Remove or filter out the missing data

In [32]:
# We see that df_nan contains string values for product_names
# Run df_prods.shape to compare number of rows in current df w number of subset once missing rows have been removed.

df_prods.shape

(49693, 5)

In [33]:
# Create new dataframe 

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [34]:
# Run following to check number of rows have decreased.

df_prods_clean.shape

(49677, 5)

In [49]:
# Another way to drop all missing values:
    # df_prods.dropna(inplace == True)

# To drop only NaNs from specific column:
    # df_prods.drop(subset = ['product'], inplace == True)

#Both methods can be risky, unless you're absolutely sure it's safe to drop values in question

In [36]:
## Duplicates
# Before culling duplicates, important to know what kind of duplicates exist
    # Ex. two rows might have same 'product_id' or multiple rows of exact values
# Record occurrences, export examples and send them to client for further explanation.

# Looking for full duplicates (rows with exact values)

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [37]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [ ]:
## Addressing duplicates

# First check the number of rows in df_prods_clean to compare number after removing duplicates.

df_prods_clean.shape

In [39]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [42]:
df_prods_clean_no_dups.shape

(49672, 5)

In [43]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

In [48]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# Tasks

In [56]:
# Task 2
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [58]:
# Everything is consistent and expected except for max prices

In [60]:
# Task 3: Checking df_ords for mixed data types
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)

In [61]:
# No column appeared with any mixed data types.

In [62]:
# Task 4: Fixing any mixed-type data. No columns to fix.

In [63]:
# Task 5: Check for missing values
df_ords.isnull().sum()

Unnamed: 0                       0
order_id                         0
user_id                          0
order_number                     0
order_days_of_week               0
order_hour_of_day                0
days_from_previous_order    206209
orer_hour_of_day                 0
dtype: int64

In [76]:
df_ords_nan = df_ords[df_ords['days_from_previous_order'].isnull() == True]

In [77]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  order_days_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_from_previous_order  orer_hour_of_day  
0                        8                       NaN                 8  
11                      11                       NaN                11  
26                      14                       NaN                14  
39                      11                       NaN                11  
45                      12                       NaN                12  
...                    ...                       ...               ...  
3420930                 12                       NaN                12  
3420934                 18                       NaN                18  
3421002                 19                       NaN                19  
3421019                 15                       NaN                15  
3421069                 11                       NaN                11  

[206209 rows x 8 columns]

In [66]:
df_ords_nan.describe()

Unnamed: 0      order_id        user_id  order_number  \
count  2.062090e+05  2.062090e+05  206209.000000      206209.0   
mean   1.712637e+06  1.708462e+06  103105.000000           1.0   
std    9.874816e+05  9.881299e+05   59527.555167           0.0   
min    0.000000e+00  2.000000e+01       1.000000           1.0   
25%    8.579730e+05  8.507300e+05   51553.000000           1.0   
50%    1.717068e+06  1.706246e+06  103105.000000           1.0   
75%    2.570518e+06  2.564292e+06  154657.000000           1.0   
max    3.421069e+06  3.421081e+06  206209.000000           1.0   

       order_days_of_week  order_hour_of_day  days_from_previous_order  \
count       206209.000000      206209.000000                       0.0   
mean             2.754118          13.626597                       NaN   
std              2.076205           4.223769                       NaN   
min              0.000000           0.000000                       NaN   
25%              1.000000          11.000000                       NaN   
50%              3.000000          14.000000                       NaN   
75%              5.000000          17.000000                       NaN   
max              6.000000          23.000000                       NaN   

       orer_hour_of_day  
count     206209.000000  
mean          13.626597  
std            4.223769  
min            0.000000  
25%           11.000000  
50%           14.000000  
75%           17.000000  
max           23.000000

1) There are 206,209 missing values from 'days_from_previous_order' column.

2) Looking at descriptive states for the column, the results are unordinary.

3) Since the column is 'days_from_previous_order' reason could be because these customers have not ordered anything else besides their initial order.

4) I would create a flag variable such as 'first_time_customer' to link to these orders so we can keep data rather than omitting. 

In [78]:
# Task 6: Addressing missing values using appropriate method
df_ords_clean = df_ords

In [79]:
# Create new variable (new_customer) to flag missing variables
df_ords_clean['new_customer'] = df_ords['days_from_previous_order'].isnull() == True

In [80]:
df_ords_clean.head()

Unnamed: 0  order_id  user_id  order_number  order_days_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_from_previous_order  orer_hour_of_day  new_customer  
0                  8                       NaN                 8          True  
1                  7                      15.0                 7         False  
2                 12                      21.0                12         False  
3                  7                      29.0                 7         False  
4                 15                      28.0                15         False

In [84]:
# Task 7: Checking for duplicate values
df_ord_dups = df_ords_clean[df_ords_clean.duplicated()]

In [85]:
df_ord_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_days_of_week, order_hour_of_day, days_from_previous_order, orer_hour_of_day, new_customer]
Index: []

In [86]:
# Task 8: No duplicates to address.

In [87]:
# Task 9: Export
df_ord_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))